Replace `clipboard.js` with native Clipboard API

ciampo opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
<!-- Please describe what you have changed or added -->

Closes #37700

Changes in this PR:
- rewrite the `useCopyToClipboard` and the (deprecated) `useCopyOnClick` hooks in `@wordpress/compose` so that they use the native [Clipboard API](https://w3c.github.io/clipboard-apis/#navigator-clipboard)
- remove the `clipboard` dependency from the project's list of dependencies (saving `2.55 kB (23%)` for the `@wordpress/compose`  package)
- add unit tests for both hooks

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

- Project builds correctly
- Test pass (including the ones added in this PR)
- Usages of the `useCopyToClipboard` hook work as expected (same as production) — see screenshots below for a few examples

## Screenshots <!-- if applicable -->

"Copy" menu item when editing any block in the editor:

https://user-images.githubusercontent.com/1083581/148126016-fa7dfa8b-2bd5-4f39-a14d-909e478fa830.mp4

"Copy URL to clipboard" button when viewing an attachment details in the media library:

![image](https://user-images.githubusercontent.com/1083581/148126072-66cd2e6d-e09f-418e-a7cf-0e18a659279b.png)

"Copy URL" button when editing an attachment in the File block:

![image](https://user-images.githubusercontent.com/1083581/148126312-3d0e59ff-6344-405a-9d56-391268bccf19.png)

"Copy" button when reviewing a post's info after publishing it:

![image](https://user-images.githubusercontent.com/1083581/148126296-71ce9730-b3a5-4ce7-ba53-ee5564e34e10.png)


## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- N/A I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

ciampo committed:
Refactor `useCopyToClipboard` hook to use the native `clipboard` APIs

ciampo committed:
Refactor `useCopyOnClick` hook to use the native `clipboard` APIs

ciampo committed:
Remove unused `clipboard` dependency

ciampo committed:
Add `useCopyToClipboard` unit tests

ciampo committed:
Add `useCopyOnClick` tests

ciampo commented on packages/compose/src/hooks/use-copy-on-click/index.js:
```diff
@@ -30,45 +25,67 @@ export default function useCopyOnClick( ref, text, timeout = 4000 ) {
+			currentWindow?.navigator?.clipboard
+				?.writeText( textToCopy )
+				.then( () => {
+					// Clear selection and ensure focus stays on the trigger element.
+					if ( trigger ) {
+						currentWindow?.getSelection()?.removeAllRanges();
+
+						trigger.focus();
+					}
```
The previous `clipboard.js` implementation used a hack to copy text to the clipboard which involved creating a temporary element, moving focus to it, selecting its text, and executing a "copy" command.

By switching to the new Clipboard APIs, clearing the selection and restoring focus may not be necessary — what do folks think?

noahtallen replied:
I agree. If that's the case, we shouldn't need the focus handling. It's probably worth double checking that focus stays on the button after clicking it still :)

ciampo commented on packages/compose/src/hooks/use-copy-on-click/index.js:
```diff
@@ -30,45 +25,67 @@ export default function useCopyOnClick( ref, text, timeout = 4000 ) {
-		clipboard.current = new Clipboard( ref.current, {
-			text: () => ( typeof text === 'function' ? text() : text ),
-		} );
+		// `triggers` is always an array, regardless of the value of the `ref` param.
+		if ( typeof ref.current === 'string' ) {
+			// expect `ref` to be a DOM selector
+			triggers = Array.from( document.querySelectorAll( ref.current ) );
+		} else if ( 'length' in ref.current ) {
+			// Expect `ref` to be a `NodeList`
+			triggers = Array.from( ref.current );
+		} else {
+			// Expect `ref` to be a single `Element`
+			triggers = [ ref.current ];
+		}
```
Previously, the different values for `ref.current` were handles by `Clipboard`'s constructor

ciampo commented on packages/compose/src/hooks/use-copy-to-clipboard/index.js:
```diff
@@ -39,32 +34,34 @@ export default function useCopyToClipboard( text, onSuccess ) {
+			currentWindow?.navigator?.clipboard
+				?.writeText( textToCopy )
+				.then( () => {
+					// Clear selection and ensure focus stays on the trigger element.
+					currentWindow?.getSelection()?.removeAllRanges();
+					trigger.focus();
```
Same question as in [this comment](https://github.com/WordPress/gutenberg/pull/37713/files?diff=split&w=0#r778404668)

ciampo commented:
Not sure how to fix the Static Analysis CI errors — running `npm install` or `npx lerna bootstrap` doesn't produce any further file changes

noahtallen commented:
> Not sure how to fix the Static Analysis CI errors — running npm install

Taking a wild guess -- is the npm version the same?

noahtallen commented on packages/compose/src/hooks/use-copy-on-click/index.js:
```diff
@@ -30,45 +25,67 @@ export default function useCopyOnClick( ref, text, timeout = 4000 ) {
 	useEffect( () => {
 		/** @type {number | undefined} */
 		let timeoutId;
+		/** @type Array<Element>) */
```
Out of curiosity, does the `compose` package not support TypeScript directly? I haven't seen this syntax before :) Is it jsdoc?

ZebulanStanphill replied:
https://www.typescriptlang.org/docs/handbook/intro-to-js-ts.html

ciampo replied:
As Zebulan pointed out, these are type hints for the TypeScript server written with the JSDoc syntax — I decided to adopt this syntax to match other files in the `@wordpress/compose` package — although we should definitely consider migrating the whole package to TypeScript in the future.

noahtallen commented on packages/compose/src/hooks/use-copy-on-click/index.js:
```diff
@@ -1,12 +1,7 @@
-/**
- * External dependencies
- */
-import Clipboard from 'clipboard';
```
While the changes in this file are pretty convoluted, it is a deprecated hook after all. The changes in the other hook are a lot more palatable. :P I still think it's worth it to remove the dependency.

ciampo replied:
I had the same train of thoughts, and came to the same conclusion — refactoring this deprecated hook allows us to remove the dependency from the repo :)

ciampo committed:
Remove selection clearance and focus management

ciampo committed:
CHANGELOG

ciampo committed:
Update package-loc.json

ciampo commented:
> Taking a wild guess -- is the npm version the same?

I noticed that [the `.nvmrc` file only specifies version `14`](https://github.com/WordPress/gutenberg/blob/trunk/.nvmrc), and so when running `nvm use`, since my locally installed  node version was an older `14.x.x`, it didn't update. Then I saw [this commit](https://github.com/WordPress/gutenberg/commit/1714d93487a3301b922f12f425cbc87ce7d7a252), installed specific version `14.18.1` and finally managed to solve the issue.

Anyway, I've also addressed the feedback and added a changelog entry.

noahtallen reviewed (approved):
Tests well for me :+1:

jorgefilipecosta commented:
Hi @ciampo it seems currently on the trunk the "Copy all content" button does not work at all (nothing happens when clicked, the button is the top options menu). The button to copy a block on the block elipsis menu also is not working. If I revert this PR it starts working again.

ciampo commented:
> Hi @ciampo it seems currently on the trunk the "Copy all content" button does not work at all (nothing happens when clicked, the button is the top options menu). The button to copy a block on the block elipsis menu also is not working. If I revert this PR it starts working again.

Hey @jorgefilipecosta , thank you for reporting this! Feel free to revert this PR.

Could also open an issue that shows how to reproduce the bug you just described? I'll look into it next week. Thank you!
